src/coreclr/src/jit/morph.cpp

```diff
@@ -6688,12 +6693,217 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee, const char** failReason)
 
         if (arg->isStruct)
         {
-            // Byref struct arguments are not allowed to fast tail call as the information
-            // of the caller's stack is lost when the callee is compiled.
             if (arg->passedByRef)
             {
+                // Generally a byref arg will block tail calling, as we have to
+                // make a local copy of the struct for the callee.
                 hasByrefParameter = true;
-                break;
+
+                // If we're optimizing, we may be able to pass our caller's byref to our callee,
+                // and so still be able to tail call.
+                if (opts.OptimizationEnabled())
+                {
+                    // First, see if this arg is a byref param.
+                    GenTreeLclVarCommon* const lcl = arg->GetNode()->IsImplicitByrefParameterValue(this);
+
+                    // If so, it must not be promoted; if we've promoted, then the arg will be
+                    // a local struct assembled from the promoted fields.
+                    if (lcl != nullptr)
+                    {
+                        JITDUMP("Arg [%06u] is unpromoted implicit byref V%02u, seeing if we can still tail call\n",
+                                dspTreeID(arg->GetNode()), lcl->GetLclNum());
+
+                        LclVarDsc* const varDsc = lvaGetDesc(lcl);
+
+                        if (varDsc->lvPromoted)
+                        {
+                            JITDUMP(" ... no, V%02u was promoted\n", lcl->GetLclNum());
+                        }
+                        else
+                        {
+                            // We also have to worry about introducing aliases if we bypass copying
+                            // the struct at the call. We'll do some limited analysis to see if we
+                            // can rule this out.
+                            const unsigned argLimit = 6;
+
+                            // If this is the only appearance of the byref in the method, then
+                            // aliasing is not possible.
+                            //
+                            if (varDsc->lvRefCnt(RCS_EARLY) == 1)
+                            {
+                                JITDUMP("... yes, arg is the only appearance of V%02u\n", lcl->GetLclNum());
+                                hasByrefParameter = false;
+                            }
+                            // If no other call arg refers to this byref, and no other arg is
```

Comment:
I can't find it in  https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/clr-jit-coding-conventions.md#73-commenting-code-blocks, but my personal opinion is that we should not allow comments after `}` and before 'else {`.